### ISSUE TITLE
dask.dataframe.read_csv failing where pandas.read_csv works

### ISSUE THREAD

**Issue by ijstokes** [dask/dask#1497]
I can't figure out what I'm doing wrong that one of these works while the other fails:

```python
import pandas as pd
import dask.dataframe as dd
import bz2
from dask.diagnostics import ProgressBar, Profiler, ResourceProfiler, visualize

ProgressBar().register()

pd.set_option('display.max_rows', 20)
MAX_ROWS = int(5e5)

cols = ['year', 'month', 'day_of_month', 'day_of_week', 'deptime', 'crs_deptime', 'arrtime', 
        'crs_arrtime', 'unique_carrier', 'flight_num', 'tail_num', 'actual_elapsed_time',
        'crs_elapsed_time', 'air_time', 'arrdelay', 'depdelay', 'origin', 'dest', 'distance', 
        'taxi_in', 'taxi_out', 'cancelled', 'cancellation_code', 'diverted', 'carrier_delay',
        'weather_delay', 'nas_delay', 'security_delay', 'late_aircraft_delay']

dtypes = {'cancellation_code': object, 'taxi_in': float, 'taxi_out': float, 'cancelled': bool,
          'diverted': bool, 'carrier_delay': float, 'weather_delay': float, 'nas_delay': float,
          'security_delay': float, 'late_aircraft_delay': float, 'tail_num': object,
          'crs_deptime': float, 'crs_arrtime': float, 'flight_num': float, 'crs_elapsed_time': float,
          'distance': float}

pdata = pd.read_csv('data/2000.csv', nrows=MAX_ROWS, header=0, names=cols, dtype=dtypes) # works
ddata = dd.read_csv('data/2000.csv', nrows=MAX_ROWS, header=0, names=cols, dtype=dtypes) # fails
```

The stack trace is below.  The data comes from http://stat-computing.org/dataexpo/2009/2000.csv.bz2

Here are my package versions:

```bash
$ conda list | grep -e numexpr -e pandas -e dask -e numpy                                                                                  
dask                      0.11.0                   py27_0    defaults                                                                                                                
numexpr                   2.6.1               np111py27_0    defaults                                                                                                                
numpy                     1.11.1                   py27_0    defaults                                                                                                                
pandas                    0.18.1              np111py27_0    defaults 
```

Stack trace:

```python
ValueErrorTraceback (most recent call last)
<ipython-input-3-c5a782a4debe> in <module>()
----> 1 ddata = dd.read_csv('data/2000.csv', nrows=MAX_ROWS, header=0, names=cols, dtype=dtypes) # fails

/projects/tanya/OnTime/envs/default/lib/python2.7/site-packages/dask/dataframe/csv.pyc in read_csv(urlpath, blocksize, chunkbytes, collection, lineterminator, compression, sample, enforce, storage_options, **kwargs)
    231     else:
    232         header = sample.split(b_lineterminator)[0] + b_lineterminator
--> 233     head = pd.read_csv(BytesIO(sample), **kwargs)
    234 
    235     df = read_csv_from_bytes(values, header, head, kwargs,

/projects/tanya/OnTime/envs/default/lib/python2.7/site-packages/pandas/io/parsers.pyc in parser_f(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision)
    560                     skip_blank_lines=skip_blank_lines)
    561 
--> 562         return _read(filepath_or_buffer, kwds)
    563 
    564     parser_f.__name__ = name

/projects/tanya/OnTime/envs/default/lib/python2.7/site-packages/pandas/io/parsers.pyc in _read(filepath_or_buffer, kwds)
    319                                   " together yet.")
    320     elif nrows is not None:
--> 321         return parser.read(nrows)
    322     elif chunksize or iterator:
    323         return parser

/projects/tanya/OnTime/envs/default/lib/python2.7/site-packages/pandas/io/parsers.pyc in read(self, nrows)
    813                 raise ValueError('skip_footer not supported for iteration')
    814 
--> 815         ret = self._engine.read(nrows)
    816 
    817         if self.options.get('as_recarray'):

/projects/tanya/OnTime/envs/default/lib/python2.7/site-packages/pandas/io/parsers.pyc in read(self, nrows)
   1312     def read(self, nrows=None):
   1313         try:
-> 1314             data = self._reader.read(nrows)
   1315         except StopIteration:
   1316             if self._first_chunk:

pandas/parser.pyx in pandas.parser.TextReader.read (pandas/parser.c:8748)()

pandas/parser.pyx in pandas.parser.TextReader._read_low_memory (pandas/parser.c:9208)()

pandas/parser.pyx in pandas.parser.TextReader._read_rows (pandas/parser.c:10022)()

pandas/parser.pyx in pandas.parser.TextReader._convert_column_data (pandas/parser.c:11397)()

pandas/parser.pyx in pandas.parser.TextReader._convert_tokens (pandas/parser.c:12775)()

ValueError: cannot safely convert passed user dtype of |b1 for int64 dtyped data in column 21
```

**Comment by ijstokes** [comment]
```
$ cut -f 22 -d , data/2000.csv | sort | uniq -c | sort -g
      1 Cancelled
 187490 1
5495557 0
```

**Comment by mrocklin** [comment]
This may be resolved by https://github.com/dask/dask/pull/1495

**Comment by ijstokes** [comment]
Splitting the file into 56 smaller pieces works.

```
split -d -l 100000 2000.csv iata_2000_
for f in iata_2000_*; do mv $f $f.csv; done
```

Then change the `dask` line above to be:

```
ddata = dd.read_csv('data/iata_2000_*.csv', nrows=MAX_ROWS, header=0, names=cols, dtype=dtypes) # works
```

**Comment by elisarver** [comment]
Yes, this looks like a float that showed up at the end of the buffer. it's one of the cases I described in #1495 as another possible parse error when a line is cut off at the end of the sample buffer.

if you temporarily add `sample=250,001` and it starts working again, then #1495 will fix it.

**Comment by jcrist** [comment]
This is fixed by #1495:

```python
In [1]: %load test.py

In [2]: # %load test.py
import dask.dataframe as dd
import pandas as pd


MAX_ROWS = int(5e5)

cols = ['year', 'month', 'day_of_month', 'day_of_week', 'deptime',
        'crs_deptime', 'arrtime', 'crs_arrtime', 'unique_carrier',
        'flight_num', 'tail_num', 'actual_elapsed_time', 'crs_elapsed_time',
        'air_time', 'arrdelay', 'depdelay', 'origin', 'dest', 'distance',
        'taxi_in', 'taxi_out', 'cancelled', 'cancellation_code', 'diverted',
        'carrier_delay', 'weather_delay', 'nas_delay', 'security_delay',
        'late_aircraft_delay']

dtypes = {'cancellation_code': object, 'taxi_in': float, 'taxi_out': float,
          'cancelled': bool, 'diverted': bool, 'carrier_delay': float,
          'weather_delay': float, 'nas_delay': float, 'security_delay': float,
          'late_aircraft_delay': float, 'tail_num': object, 'crs_deptime':
          float, 'crs_arrtime': float, 'flight_num': float, 'crs_elapsed_time':
          float, 'distance': float}

ddata = dd.read_csv('2000.csv', nrows=MAX_ROWS, header=0, names=cols,
                    dtype=dtypes)
   ...:

In [3]: len(ddata)
Out[3]: 500000
```

Closing.
